core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+    private List<MetadataUpdate> changes = Lists.newArrayList();
+
+    private Builder() {}
+
+    private Builder(ViewMetadata base) {
+      this.formatVersion = base.formatVersion();
+      this.location = base.location();
+      this.currentSchemaId = base.currentSchemaId();
+      this.schemas = Lists.newArrayList(base.schemas());
+      this.currentVersionId = base.currentVersionId();
+      this.versions = Lists.newArrayList(base.versions());
+      this.history = Lists.newArrayList(base.history());
+      this.properties = Maps.newHashMap(base.properties());
+      this.changes = Lists.newArrayList(base.changes());
+    }
+
+    public Builder upgradeFormatVersion(int newFormatVersion) {
+      Preconditions.checkArgument(
+          newFormatVersion >= this.formatVersion,
+          "Cannot downgrade v%s view to v%s",
+          formatVersion,
+          newFormatVersion);
+
+      if (this.formatVersion == newFormatVersion) {
+        return this;
+      }
+
+      this.formatVersion = newFormatVersion;
+      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      return this;
+    }
+
+    public Builder setLocation(String newLocation) {
+      if (null != this.location && this.location.equals(newLocation)) {
+        return this;
+      }
+
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder setCurrentVersionId(int newVersionId) {
+      if (this.currentVersionId == newVersionId) {
+        return this;
+      }
+
+      this.currentVersionId = newVersionId;
+      this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+      return this;
+    }
+
+    public Builder setCurrentSchemaId(Integer newSchemaId) {
+      if (null != this.currentSchemaId && this.currentSchemaId.equals(newSchemaId)) {
+        return this;
+      }
+
+      this.currentSchemaId = newSchemaId;
+      this.changes.add(new MetadataUpdate.SetCurrentSchema(newSchemaId));
+      return this;
+    }
+
+    public Builder addSchema(Schema schema) {
+      this.schemas.add(schema);
+      this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      return this;
+    }
+
+    public Builder setSchemas(Iterable<Schema> schemasToAdd) {
+      for (Schema schema : schemasToAdd) {
+        this.schemas.add(schema);
+        this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      }
+
+      return this;
+    }
+
+    public Builder addVersion(ViewVersion version) {
+      this.versions.add(version);
+      this.changes.add(new MetadataUpdate.AddViewVersion(version));
+      this.history.add(
+          ImmutableViewHistoryEntry.builder()
+              .timestampMillis(version.timestampMillis())
+              .versionId(version.versionId())
+              .build());
+      return this;
+    }
+
+    public Builder setVersions(Iterable<ViewVersion> versionsToAdd) {
```

Comment:
Looks like this isn't named correctly. It isn't clear what "set" would mean. Luckily, I don't think we need a method to add multiple versions at once, do we?

Reply:
previously we were mainly using it in `ViewMetadataParser`, but that's not the case anymore, so I've removed it for now.